### DISCUSSION TITLE
Option to ignore forks

### DISCUSSION THREAD

**Discussion by heaths** [community/community#13126] [Other Feature Feedback, Questions, & Ideas] [answered]
When searching for code, it would be preferable to see authoritative results only - to ignore forks which can also cause needles scrolling through pages of results.

For example, this [search](https://cs.github.com/?scopeName=All+repos&scope=&q=ChallengeBasedAuthenticationPolicy) shows a lot of forks. They might help, but may also be far behind the main repo from which they were forked.

GitHub already knows if a repo is a fork so I'm hoping that can permeate to search results and we can exclude them. Or maybe even exclude them by default and let searchers opt into seeing them. I imagine there's good reasons to search forks, but doesn't seem helpful by default.

**Comment by bagder** [comment]
Plain forks would be easy for GitHub to ignore and it would be **most** useful.

I do however have an extended problem, and that is that I author a library (curl/curl) that hundreds of other projects **import** into their repositories. If I try to find users of this library, like for example if I want to check if I can find applications that use a particular option, there are literally **thousands** of forks and code imports that make it next to impossible to find what I'm looking for.

Test case: [Look for CURLOPT_HEADERFUNCTION](https://github.com/search?q=CURLOPT_HEADERFUNCTION&type=code)

**Reply by look** [reply]
Here's that same query with [the new code search](https://cs.github.com/?scopeName=All+repos&scope=&q=CURLOPT_HEADERFUNCTION).

What kind of results or filters are you looking for to meet your use case? (I'm not familiar with the code in the results so it's hard for me to judge the quality.)

**Reply by jsoref** [reply]
Here's an example:
https://cs.github.com/?scopeName=All+repos&scope=&q=%22curl_thread_t+Curl_thread_create%28unsigned+int+%28*func%29+%28void+*%29%2C+void+*arg%29%22

Consider these hits:

> nrfconnect/sdk-nrf
‎ext/curl/lib/curl_threads.c‎
1 match
C
main

```
  return 0;
}

curl_thread_t Curl_thread_create(unsigned int (*func) (void *), void *arg)
{
  curl_thread_t t = malloc(sizeof(pthread_t));
  struct Curl_actual_call *ac = malloc(sizeof(struct Curl_actual_call));
```

https://github.com/nrfconnect/sdk-nrf/blob/b6f76303347d8d5cf37c4186b4738cbce41b20c1/ext/curl/lib/curl_threads.c

It's a copy of [curl's](https://github.com/curl/curl/blob/5121d98c22dc077d72c68e7b9436a04f55bd7249/lib/curl_threads.c) -- not sure the specific version, and I'm not going to try.

> whitecatboard/Lua-RTOS-ESP32
‎components/curl/lib/curl_threads.c‎
1 match
C
master

```
  return 0;
}

curl_thread_t Curl_thread_create(unsigned int (*func) (void *), void *arg)
{
  curl_thread_t t = malloc(sizeof(pthread_t));
  struct curl_actual_call *ac = malloc(sizeof(struct curl_actual_call));
```

Ditto.

The point is that one doesn't want to see a committed copy of the code, just consumer code (which will be in the same repo, since, after all, they vendored the code, so presumably, they're making some calls to the library).

**Reply by jsoref** [reply]
But, even given your example, or rather, a [variant](https://cs.github.com/?q=CURLOPT_HEADERFUNCTION+NOT+%22curl%2C+CURLOPT_HEADERFUNCTION%22&p=4&pt=0721c5461381c6a2aea00d480eb8f128f9727a96fd9591fa475b23cdd37abfc99df298e15ee3296a2f59665dfc0ba6f70abb99e0267725a644ecaa375794495dca088156d6efa2fc80fea37f0416ea28fe7a8d94cd52b981f94ec32bb453b5bb63e6dfa411865745c7bd4c4234548a9c095e958f5bba4bdafcd1d1f06aba4cd22f845c06c149d38cf5260f18580329ba06f4bd0f1fc29adf5dfac9de9d6a3778b77b111a7eef24b27bcad56a98d37cb223a72b02e4aad047f8b7da19641efe004a0e7e83a34029ff2bfe43db92ce778d899408e017f0fcaaa40446919e31a0a366725fc3f1bfd8d80e696f759f983a56b0c56b7d24f38e0ae3a7ba040cf570af1f12150c34528e6cd8697261c51bd29f27ddff987c70be81f888c1dce116a5b6665133fac2bbfb303f2112d8085b3a2f57d2438269a75e315b19f896018bb887a3693eda7dc5c454b597becbd572f019feb4517f94b906d7d3520141be2f2f89c680534a258a2868bda3ccdd3f0bde2574e9bf5c092d578d8b404ba69ee1e6ce4609f4fd8eba5fd0f7938094494085ad&scope=&scopeName=All+repos)

Here's one where someone uploaded a copy of `curl` but didn't tell github it was a fork of curl/curl:

> [MikeRalphson/curl](https://cs.github.com/MikeRalphson/curl?q=CURLOPT_HEADERFUNCTION%20NOT%20%22curl%2C%20CURLOPT_HEADERFUNCTION%22)
[‎lib/url.c‎](https://cs.github.com/MikeRalphson/curl/blob/cba16bf38dc0fe40d65b5f4e266fbf96de014637/lib/url.c?q=CURLOPT_HEADERFUNCTION+NOT+%22curl%2C+CURLOPT_HEADERFUNCTION%22)
1 match
C
master

```c
    if(!data->set.err)
      data->set.err = stderr;
    break;
  case CURLOPT_HEADERFUNCTION:
    /*
     * Set header write callback
     */
```

https://github.com/MikeRalphson/curl/blob/cba16bf38dc0fe40d65b5f4e266fbf96de014637/lib/url.c

**Comment by look** [comment]
@heaths Thanks for the feedback. This is on our roadmap, so stay tuned!

**Reply by scottaddie** [reply]
Came here to provide the same feedback and see that someone on my own team (@heaths) reported it already. 😄

**Reply by Dentrax** [reply]
Any update on this?

**Reply by look** [reply]
Still on our roadmap. We're currently focused on scaling the index to full GitHub scale. Once that's done, we can tackle this.

**Reply by Dentrax** [reply]
I think it'd be better to ignore forks _by default_ for better UX. We can pass a `forks` bool to enable listing forks. Wdyt?

**Reply by kidonng** [reply]
I don't know about the technical details or how many forks are out there so I may be wrong, but implementing fork filtering should help scaling as you can put off the indexing work for forks.

@Dentrax Ignoring forks is definitely going to be the default, it's just not implemented. Also, better use the [existing syntax](https://docs.github.com/search-github/searching-on-github/searching-in-forks) for regular searches `fork:true` / `fork:only`.

**Reply by nickserv** [reply]
For reference, Sourcegraph automatic alt ignores archives and forks from GitHub by default. Both can be included with Boolean flags.

**Reply by Dentrax** [reply]
Kindly ping @look. I just wanted to circle back on this ask.

**Reply by look** [reply]
We still plan to do this, but have been focusing on building out the system before adding additional filters to the index.

**Reply by nicowilliams** [reply]
This is the single biggest problem with GitHub search _by far_.  There is so so much noise from forks in any search for any common thing.  Want to know who uses some OpenSSL function?  No, you can't get the answer, because you'll see the thousands of forks of OpenSSL first, with occasional hits mixed in, so you can't even skip to page 500 or something to begin finding non-fork hits.  Basically, GitHub search is useless to me.  Like right now I'm looking for "tpm2" in code written in Rust, and as esoteric as one might think that search might be... nope, the search results are full of noise from forks.

**Comment by Trinhan87** [comment]
> When searching for code, it would be preferable to see authoritative results only - to ignore forks which can also cause needles scrolling through pages of results.
> 
> For example, this [search](https://cs.github.com/?scopeName=All+repos&scope=&q=ChallengeBasedAuthenticationPolicy) shows a lot of forks. They might help, but may also be far behind the main repo from which they were forked.
> 
> GitHub already knows if a repo is a fork so I'm hoping that can permeate to search results and we can exclude them. Or maybe even exclude them by default and let searchers opt into seeing them. I imagine there's good reasons to search forks, but doesn't seem helpful by default.

**Comment by look** [comment, marked as answer]
Thanks for your patience, everyone. There is now an option to exclude forks (or limit your search to forks).

To exclude forks from your results:

```
NOT is:fork
```

To search only fork repositories:

```
is:fork
```

For more information, please [see the documentation](https://docs.github.com/en/search-github/github-code-search/understanding-github-code-search-syntax#is-qualifier).

**Reply by jsoref** [reply]
Thanks.

It'd be nice if non owners of repositories could crowdsource a tag for `unregistered-fork` as exemplified in https://github.com/orgs/community/discussions/13126#discussioncomment-2909457.

As a given repository intentionally bypassed the `fork` button, we can't rely on the owners to add a tag...

Do you have any suggestions for how to address this edge / plans for it going forward?

**Reply by look** [reply]
For exact duplicate files, our search engine should already pick the "best" repository to show, which should be the original (or a more highly ranked fork in the case when the original repository has been superseded).

That doesn't cover files that have changed, either in the original repository or the fork. For these, we are looking into adding document similarity and improving our ranking.

In the meantime, I recommend a save scope populated with repositories you want to exclude.

**Comment by monperrus** [comment]
I confirm that `NOT is:fork` works as expected

https://github.com/search?q=%22Decompiler+and+security+analysis+tool+for+all+deployed+contracts%22+NOT+is%3Afork&type=code

**Comment by Func86** [comment]
The [advanced search page](https://github.com/search/advanced) can't generate the correct syntax for code search, it uses `fork:true` and `fork:only`, which are for repo search.
> Possible unrecognized qualifier, searching for this term literally

**Comment by dwallman1-pixel** [comment]
> When searching for code, it would be preferable to see authoritative results only - to ignore forks which can also cause needles scrolling through pages of results.
> 
> For example, this [search](https://cs.github.com/?scopeName=All+repos&scope=&q=ChallengeBasedAuthenticationPolicy) shows a lot of forks. They might help, but may also be far behind the main repo from which they were forked.
> 
> GitHub already knows if a repo is a fork so I'm hoping that can permeate to search results and we can exclude them. Or maybe even exclude them by default and let searchers opt into seeing them. I imagine there's good reasons to search forks, but doesn't seem helpful by default.

https://github.com/cli/cli/issues/9598
